File: spine-cocos2dx/3.0/src/spine/SkeletonRenderer.cpp
/******************************************************************************
 * Spine Runtimes Software License
 * Version 2.1
 * 
 * Copyright (c) 2013, Esoteric Software
 * All rights reserved.
 * 
 * You are granted a perpetual, non-exclusive, non-sublicensable and
 * non-transferable license to install, execute and perform the Spine Runtimes
 * Software (the "Software") solely for internal use. Without the written
 * permission of Esoteric Software (typically granted by licensing Spine), you
 * may not (a) modify, translate, adapt or otherwise create derivative works,
 * improvements of the Software or develop new applications using the Software
 * or (b) remove, delete, alter or obscure any trademarks or any copyright,
 * trademark, patent or other intellectual property or proprietary rights
 * notices on or in the Software, including any copy thereof. Redistributions
 * in binary or source form must include this license and terms.
 * 
 * THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL ESOTERIC SOFTARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

#include <spine/SkeletonRenderer.h>
#include <spine/spine-cocos2dx.h>
#include <spine/extension.h>
#include <spine/PolygonBatch.h>
#include <algorithm>

USING_NS_CC;
using std::min;
using std::max;

namespace spine {

static const int quadTriangles[6] = {0, 1, 2, 2, 3, 0};

SkeletonRenderer* SkeletonRenderer::createWithData (spSkeletonData* skeletonData, bool ownsSkeletonData) {
	SkeletonRenderer* node = new SkeletonRenderer(skeletonData, ownsSkeletonData);
	node->autorelease();
	return node;
}

SkeletonRenderer* SkeletonRenderer::createWithFile (const char* skeletonDataFile, spAtlas* atlas, float scale) {
	SkeletonRenderer* node = new SkeletonRenderer(skeletonDataFile, atlas, scale);
	node->autorelease();
	return node;
}

SkeletonRenderer* SkeletonRenderer::createWithFile (const char* skeletonDataFile, const char* atlasFile, float scale) {
	SkeletonRenderer* node = new SkeletonRenderer(skeletonDataFile, atlasFile, scale);
	node->autorelease();
	return node;
}

void SkeletonRenderer::initialize () {
	atlas = 0;
	debugSlots = false;
	debugBones = false;
	timeScale = 1;

	worldVertices = MALLOC(float, 1000); // Max number of vertices per mesh.

	batch = PolygonBatch::createWithCapacity(2000); // Max number of vertices and triangles per batch.
	batch->retain();

	blendFunc = BlendFunc::ALPHA_PREMULTIPLIED;
	setOpacityModifyRGB(true);

	setShaderProgram(ShaderCache::getInstance()->getProgram(GLProgram::SHADER_NAME_POSITION_TEXTURE_COLOR));
	scheduleUpdate();
}

void SkeletonRenderer::setSkeletonData (spSkeletonData *skeletonData, bool ownsSkeletonData) {
	skeleton = spSkeleton_create(skeletonData);
	rootBone = skeleton->bones[0];
	this->ownsSkeletonData = ownsSkeletonData;	
}

SkeletonRenderer::SkeletonRenderer () {
	initialize();
}

SkeletonRenderer::SkeletonRenderer (spSkeletonData *skeletonData, bool ownsSkeletonData) {
	initialize();

	setSkeletonData(skeletonData, ownsSkeletonData);
}

SkeletonRenderer::SkeletonRenderer (const char* skeletonDataFile, spAtlas* atlas, float scale) {
	initialize();

	spSkeletonJson* json = spSkeletonJson_create(atlas);
	json->scale = scale == 0 ? (1 / Director::getInstance()->getContentScaleFactor()) : scale;
	spSkeletonData* skeletonData = spSkeletonJson_readSkeletonDataFile(json, skeletonDataFile);
	CCASSERT(skeletonData, json->error ? json->error : "Error reading skeleton data.");
	spSkeletonJson_dispose(json);

	setSkeletonData(skeletonData, true);
}

SkeletonRenderer::SkeletonRenderer (const char* skeletonDataFile, const char* atlasFile, float scale) {
	initialize();

	atlas = spAtlas_createFromFile(atlasFile, 0);
	CCASSERT(atlas, "Error reading atlas file.");

	spSkeletonJson* json = spSkeletonJson_create(atlas);
	json->scale = scale == 0 ? (1 / Director::getInstance()->getContentScaleFactor()) : scale;
	spSkeletonData* skeletonData = spSkeletonJson_readSkeletonDataFile(json, skeletonDataFile);
	CCASSERT(skeletonData, json->error ? json->error : "Error reading skeleton data file.");
	spSkeletonJson_dispose(json);

	setSkeletonData(skeletonData, true);
}

SkeletonRenderer::~SkeletonRenderer () {
	if (ownsSkeletonData) spSkeletonData_dispose(skeleton->data);
	if (atlas) spAtlas_dispose(atlas);
	spSkeleton_dispose(skeleton);
	batch->release();
}

void SkeletonRenderer::update (float deltaTime) {
	spSkeleton_update(skeleton, deltaTime * timeScale);
}

void SkeletonRenderer::draw(Renderer* renderer, const kmMat4& transform, bool transformUpdated) {
    drawCommand.init(_globalZOrder);
    drawCommand.func = CC_CALLBACK_0(SkeletonRenderer::drawSkeleton, this, transform, transformUpdated);
    renderer->addCommand(&drawCommand);
}

void SkeletonRenderer::drawSkeleton (const kmMat4& transform, bool transformUpdated) {
	getShaderProgram()->use();
	getShaderProgram()->setUniformsForBuiltins(transform);

	Color3B nodeColor = getColor();
	skeleton->r = nodeColor.r / (float)255;
	skeleton->g = nodeColor.g / (float)255;
	skeleton->b = nodeColor.b / (float)255;
	skeleton->a = getOpacity() / (float)255;

	int additive = -1;
	Color4B color;
	const float* uvs = nullptr;
	int verticesCount = 0;
	const int* triangles = nullptr;
	int trianglesCount = 0;
	float r = 0, g = 0, b = 0, a = 0;
	for (int i = 0, n = skeleton->slotCount; i < n; i++) {
		spSlot* slot = skeleton->drawOrder[i];
		if (!slot->attachment) continue;
		Texture2D *texture = nullptr;
		switch (slot->attachment->type) {
		case SP_ATTACHMENT_REGION: {
			spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
			spRegionAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot->bone, worldVertices);
			texture = getTexture(attachment);
			uvs = attachment->uvs;
			verticesCount = 8;
			triangles = quadTriangles;
			trianglesCount = 6;
			r = attachment->r;
			g = attachment->g;
			b = attachment->b;
			a = attachment->a;
			break;
		}
		case SP_ATTACHMENT_MESH: {
			spMeshAttachment* attachment = (spMeshAttachment*)slot->attachment;
			spMeshAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot, worldVertices);
			texture = getTexture(attachment);
			uvs = attachment->uvs;
			verticesCount = attachment->verticesCount;
			triangles = attachment->triangles;
			trianglesCount = attachment->trianglesCount;
			r = attachment->r;
			g = attachment->g;
			b = attachment->b;
			a = attachment->a;
			break;
		}
		case SP_ATTACHMENT_SKINNED_MESH: {
			spSkinnedMeshAttachment* attachment = (spSkinnedMeshAttachment*)slot->attachment;
			spSkinnedMeshAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot, worldVertices);
			texture = getTexture(attachment);
			uvs = attachment->uvs;
			verticesCount = attachment->uvsCount;
			triangles = attachment->triangles;
			trianglesCount = attachment->trianglesCount;
			r = attachment->r;
			g = attachment->g;
			b = attachment->b;
			a = attachment->a;
			break;
		}
		} 
		if (texture) {
			if (slot->data->additiveBlending != additive) {
				batch->flush();
				GL::blendFunc(blendFunc.src, slot->data->additiveBlending ? GL_ONE : blendFunc.dst);
				additive = slot->data->additiveBlending;
			}
			color.a = skeleton->a * slot->a * a * 255;
			float multiplier = premultipliedAlpha ? color.a : 255;
			color.r = skeleton->r * slot->r * r * multiplier;
			color.g = skeleton->g * slot->g * g * multiplier;
			color.b = skeleton->b * slot->b * b * multiplier;
			batch->add(texture, worldVertices, uvs, verticesCount, triangles, trianglesCount, &color);
		}
	}
	batch->flush();

	if (debugSlots || debugBones) {
		kmGLPushMatrix();
		kmGLLoadMatrix(&transform);
		if (debugSlots) {
			// Slots.
			ccDrawColor4B(0, 0, 255, 255);
			glLineWidth(1);
			CCPoint points[4];
			ccV3F_C4B_T2F_Quad quad;
			for (int i = 0, n = skeleton->slotCount; i < n; i++) {
				spSlot* slot = skeleton->drawOrder[i];
				if (!slot->attachment || slot->attachment->type != SP_ATTACHMENT_REGION) continue;
				spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
				spRegionAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot->bone, worldVertices);
				points[0] = Point(worldVertices[0], worldVertices[1]);
				points[1] = Point(worldVertices[2], worldVertices[3]);
				points[2] = Point(worldVertices[4], worldVertices[5]);
				points[3] = Point(worldVertices[6], worldVertices[7]);
				ccDrawPoly(points, 4, true);
			}
		}
		if (debugBones) {
			// Bone lengths.
			glLineWidth(2);
			ccDrawColor4B(255, 0, 0, 255);
			for (int i = 0, n = skeleton->boneCount; i < n; i++) {
				spBone *bone = skeleton->bones[i];
				float x = bone->data->length * bone->m00 + bone->worldX;
				float y = bone->data->length * bone->m10 + bone->worldY;
				ccDrawLine(Point(bone->worldX, bone->worldY), Point(x, y));
			}
			// Bone origins.
			ccPointSize(4);
			ccDrawColor4B(0, 0, 255, 255); // Root bone is blue.
			for (int i = 0, n = skeleton->boneCount; i < n; i++) {
				spBone *bone = skeleton->bones[i];
				ccDrawPoint(Point(bone->worldX, bone->worldY));
				if (i == 0) ccDrawColor4B(0, 255, 0, 255);
			}
		}
		kmGLPopMatrix();
	}
}

Texture2D* SkeletonRenderer::getTexture (spRegionAttachment* attachment) const {
	return (Texture2D*)((spAtlasRegion*)attachment->rendererObject)->page->rendererObject;
}

Texture2D* SkeletonRenderer::getTexture (spMeshAttachment* attachment) const {
	return (Texture2D*)((spAtlasRegion*)attachment->rendererObject)->page->rendererObject;
}

Texture2D* SkeletonRenderer::getTexture (spSkinnedMeshAttachment* attachment) const {
	return (Texture2D*)((spAtlasRegion*)attachment->rendererObject)->page->rendererObject;
}

Rect SkeletonRenderer::boundingBox () {
	float minX = FLT_MAX, minY = FLT_MAX, maxX = FLT_MIN, maxY = FLT_MIN;
	float scaleX = getScaleX();
	float scaleY = getScaleY();
	float vertices[8];
	for (int i = 0; i < skeleton->slotCount; ++i) {
		spSlot* slot = skeleton->slots[i];
		if (!slot->attachment || slot->attachment->type != SP_ATTACHMENT_REGION) continue;
		spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
		spRegionAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot->bone, vertices);
		minX = min(minX, vertices[SP_VERTEX_X1] * scaleX);
		minY = min(minY, vertices[SP_VERTEX_Y1] * scaleY);
		maxX = max(maxX, vertices[SP_VERTEX_X1] * scaleX);
		maxY = max(maxY, vertices[SP_VERTEX_Y1] * scaleY);
		minX = min(minX, vertices[SP_VERTEX_X4] * scaleX);
		minY = min(minY, vertices[SP_VERTEX_Y4] * scaleY);
		maxX = max(maxX, vertices[SP_VERTEX_X4] * scaleX);
		maxY = max(maxY, vertices[SP_VERTEX_Y4] * scaleY);
		minX = min(minX, vertices[SP_VERTEX_X2] * scaleX);
		minY = min(minY, vertices[SP_VERTEX_Y2] * scaleY);
		maxX = max(maxX, vertices[SP_VERTEX_X2] * scaleX);
		maxY = max(maxY, vertices[SP_VERTEX_Y2] * scaleY);
		minX = min(minX, vertices[SP_VERTEX_X3] * scaleX);
		minY = min(minY, vertices[SP_VERTEX_Y3] * scaleY);
		maxX = max(maxX, vertices[SP_VERTEX_X3] * scaleX);
		maxY = max(maxY, vertices[SP_VERTEX_Y3] * scaleY);
	}
	Point position = getPosition();
	return Rect(position.x + minX, position.y + minY, maxX - minX, maxY - minY);
}

// --- Convenience methods for Skeleton_* functions.

void SkeletonRenderer::updateWorldTransform () {
	spSkeleton_updateWorldTransform(skeleton);
}

void SkeletonRenderer::setToSetupPose () {
	spSkeleton_setToSetupPose(skeleton);
}
void SkeletonRenderer::setBonesToSetupPose () {
	spSkeleton_setBonesToSetupPose(skeleton);
}
void SkeletonRenderer::setSlotsToSetupPose () {
	spSkeleton_setSlotsToSetupPose(skeleton);
}

spBone* SkeletonRenderer::findBone (const char* boneName) const {
	return spSkeleton_findBone(skeleton, boneName);
}

spSlot* SkeletonRenderer::findSlot (const char* slotName) const {
	return spSkeleton_findSlot(skeleton, slotName);
}

bool SkeletonRenderer::setSkin (const char* skinName) {
	return spSkeleton_setSkinByName(skeleton, skinName) ? true : false;
}

spAttachment* SkeletonRenderer::getAttachment (const char* slotName, const char* attachmentName) const {
	return spSkeleton_getAttachmentForSlotName(skeleton, slotName, attachmentName);
}
bool SkeletonRenderer::setAttachment (const char* slotName, const char* attachmentName) {
	return spSkeleton_setAttachment(skeleton, slotName, attachmentName) ? true : false;
}

// --- CCBlendProtocol

const BlendFunc& SkeletonRenderer::getBlendFunc () const {
    return blendFunc;
}

void SkeletonRenderer::setBlendFunc (const BlendFunc &blendFunc) {
    this->blendFunc = blendFunc;
}

void SkeletonRenderer::setOpacityModifyRGB (bool value) {
	premultipliedAlpha = value;
}

bool SkeletonRenderer::isOpacityModifyRGB () {
	return premultipliedAlpha;
}

}

File: spine-ts/canvas/src/SkeletonRenderer.ts
/******************************************************************************
 * Spine Runtimes Software License v2.5
 *
 * Copyright (c) 2013-2016, Esoteric Software
 * All rights reserved.
 *
 * You are granted a perpetual, non-exclusive, non-sublicensable, and
 * non-transferable license to use, install, execute, and perform the Spine
 * Runtimes software and derivative works solely for personal or internal
 * use. Without the written permission of Esoteric Software (see Section 2 of
 * the Spine Software License Agreement), you may not (a) modify, translate,
 * adapt, or develop new applications using the Spine Runtimes or otherwise
 * create derivative works or improvements of the Spine Runtimes or (b) remove,
 * delete, alter, or obscure any trademarks or any copyright, trademark, patent,
 * or other intellectual property or proprietary rights notices on or in the
 * Software, including any copy thereof. Redistributions in binary or source
 * form must include this license and terms.
 *
 * THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL ESOTERIC SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES, BUSINESS INTERRUPTION, OR LOSS OF
 * USE, DATA, OR PROFITS) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

module spine.canvas {
	export class SkeletonRenderer {
		static QUAD_TRIANGLES = [0, 1, 2, 2, 3, 0];
		static VERTEX_SIZE = 2 + 2 + 4;

		private ctx: CanvasRenderingContext2D;

		public triangleRendering = false;
		public debugRendering = false;
		private vertices = Utils.newFloatArray(8 * 1024);
		private tempColor = new Color();

		constructor (context: CanvasRenderingContext2D) {
			this.ctx = context;
		}

		draw (skeleton: Skeleton) {
			if (this.triangleRendering) this.drawTriangles(skeleton);
			else this.drawImages(skeleton);
		}

		private drawImages (skeleton: Skeleton) {
			let ctx = this.ctx;
			let drawOrder = skeleton.drawOrder;

			if (this.debugRendering) ctx.strokeStyle = "green";

			ctx.save();
			for (let i = 0, n = drawOrder.length; i < n; i++) {
				let slot = drawOrder[i];
				let attachment = slot.getAttachment();
				let regionAttachment: RegionAttachment = null;
				let region: TextureAtlasRegion = null;
				let image: HTMLImageElement = null;

				if (attachment instanceof RegionAttachment) {
					regionAttachment = <RegionAttachment>attachment;
					region = <TextureAtlasRegion>regionAttachment.region;
					image = (<CanvasTexture>region.texture).getImage();
				} else continue;

				let skeleton = slot.bone.skeleton;
				let skeletonColor = skeleton.color;
				let slotColor = slot.color;
				let regionColor = regionAttachment.color;
				let alpha = skeletonColor.a * slotColor.a * regionColor.a;
				let color = this.tempColor;
				color.set(skeletonColor.r * slotColor.r * regionColor.r,
					skeletonColor.g * slotColor.g * regionColor.g,
					skeletonColor.b * slotColor.b * regionColor.b,
					alpha);

				let att = <RegionAttachment>attachment;
				let bone = slot.bone;
				let w = region.width;
				let h = region.height;
				ctx.save();
				ctx.transform(bone.a, bone.c, bone.b, bone.d, bone.worldX, bone.worldY);
				ctx.translate(attachment.offset[0], attachment.offset[1]);
				ctx.rotate(attachment.rotation * Math.PI / 180);
				ctx.scale(attachment.scaleX, attachment.scaleY);
				ctx.translate(w / 2, h / 2);
				if (attachment.region.rotate) {
					let t = w;
					w = h;
					h = t;
					ctx.rotate(-Math.PI / 2);
				}
				ctx.scale(1, -1);
				ctx.translate(-w / 2, -h / 2);
				if (color.r != 1 || color.g != 1 || color.b != 1 || color.a != 1) {
					ctx.globalAlpha = color.a;
					// experimental tinting via compositing, doesn't work
					// ctx.globalCompositeOperation = "source-atop";
					// ctx.fillStyle = "rgba(" + (color.r * 255 | 0) + ", " + (color.g * 255 | 0)  + ", " + (color.b * 255 | 0) + ", " + color.a + ")";
					// ctx.fillRect(0, 0, w, h);
				}
				ctx.drawImage(image, region.x, region.y, w, h, 0, 0, w, h);
				if (this.debugRendering) ctx.strokeRect(0, 0, w, h);
				ctx.restore();
			}

			ctx.restore();
		}

		private drawTriangles (skeleton: Skeleton) {
			let blendMode: BlendMode = null;

			let vertices: ArrayLike<number> = this.vertices;
			let triangles: Array<number> = null;
			let drawOrder = skeleton.drawOrder;

			for (let i = 0, n = drawOrder.length; i < n; i++) {
				let slot = drawOrder[i];
				let attachment = slot.getAttachment();
				let texture: HTMLImageElement = null;
				let region: TextureAtlasRegion = null;
				if (attachment instanceof RegionAttachment) {
					let regionAttachment = <RegionAttachment>attachment;
					vertices = this.computeRegionVertices(slot, regionAttachment, false);
					triangles = SkeletonRenderer.QUAD_TRIANGLES;
					region = <TextureAtlasRegion>regionAttachment.region;
					texture = (<CanvasTexture>region.texture).getImage();

				} else if (attachment instanceof MeshAttachment) {
					let mesh = <MeshAttachment>attachment;
					vertices = this.computeMeshVertices(slot, mesh, false);
					triangles = mesh.triangles;
					texture = (<TextureAtlasRegion>mesh.region.renderObject).texture.getImage();
				} else continue;

				if (texture != null) {
					let slotBlendMode = slot.data.blendMode;
					if (slotBlendMode != blendMode) {
						blendMode = slotBlendMode;
					}

					let ctx = this.ctx;

					for (var j = 0; j < triangles.length; j+=3) {
						let t1 = triangles[j] * 8, t2 = triangles[j+1] * 8, t3 = triangles[j+2] * 8;

						let x0 = vertices[t1], y0 = vertices[t1 + 1], u0 = vertices[t1 + 6], v0 = vertices[t1 + 7];
						let x1 = vertices[t2], y1 = vertices[t2 + 1], u1 = vertices[t2 + 6], v1 = vertices[t2 + 7];
						let x2 = vertices[t3], y2 = vertices[t3 + 1], u2 = vertices[t3 + 6], v2 = vertices[t3 + 7];

						this.drawTriangle(texture, x0, y0, u0, v0, x1, y1, u1, v1, x2, y2, u2, v2);

						if (this.debugRendering) {
							ctx.strokeStyle = "green";
							ctx.beginPath();
							ctx.moveTo(x0, y0);
							ctx.lineTo(x1, y1);
							ctx.lineTo(x2, y2);
							ctx.lineTo(x0, y0);
							ctx.stroke();
						}
					}
				}
			}
		}

		// Adapted from http://extremelysatisfactorytotalitarianism.com/blog/?p=2120
		// Apache 2 licensed
		private drawTriangle(img: HTMLImageElement, x0: number, y0: number, u0: number, v0: number,
						x1: number, y1: number, u1: number, v1: number,
						x2: number, y2: number, u2: number, v2: number) {
			let ctx = this.ctx;

			u0 *= img.width;
			v0 *= img.height;
			u1 *= img.width;
			v1 *= img.height;
			u2 *= img.width;
			v2 *= img.height;

			ctx.beginPath();
			ctx.moveTo(x0, y0);
			ctx.lineTo(x1, y1);
			ctx.lineTo(x2, y2);
			ctx.closePath();

			x1 -= x0;
			y1 -= y0;
			x2 -= x0;
			y2 -= y0;

			u1 -= u0;
			v1 -= v0;
			u2 -= u0;
			v2 -= v0;

			var det = 1 / (u1*v2 - u2*v1),

			// linear transformation
			a = (v2*x1 - v1*x2) * det,
			b = (v2*y1 - v1*y2) * det,
			c = (u1*x2 - u2*x1) * det,
			d = (u1*y2 - u2*y1) * det,

			// translation
			e = x0 - a*u0 - c*v0,
			f = y0 - b*u0 - d*v0;

			ctx.save();
			ctx.transform(a, b, c, d, e, f);
			ctx.clip();
			ctx.drawImage(img, 0, 0);
			ctx.restore();
		}

		private computeRegionVertices(slot: Slot, region: RegionAttachment, pma: boolean) {
			let skeleton = slot.bone.skeleton;
			let skeletonColor = skeleton.color;
			let slotColor = slot.color;
			let regionColor = region.color;
			let alpha = skeletonColor.a * slotColor.a * regionColor.a;
			let multiplier = pma ? alpha : 1;
			let color = this.tempColor;
			color.set(skeletonColor.r * slotColor.r * regionColor.r * multiplier,
					  skeletonColor.g * slotColor.g * regionColor.g * multiplier,
					  skeletonColor.b * slotColor.b * regionColor.b * multiplier,
					  alpha);

			region.computeWorldVertices(slot.bone, this.vertices, 0, SkeletonRenderer.VERTEX_SIZE);

			let vertices = this.vertices;
			let uvs = region.uvs;

			vertices[RegionAttachment.C1R] = color.r;
			vertices[RegionAttachment.C1G] = color.g;
			vertices[RegionAttachment.C1B] = color.b;
			vertices[RegionAttachment.C1A] = color.a;
			vertices[RegionAttachment.U1] = uvs[0];
			vertices[RegionAttachment.V1] = uvs[1];

			vertices[RegionAttachment.C2R] = color.r;
			vertices[RegionAttachment.C2G] = color.g;
			vertices[RegionAttachment.C2B] = color.b;
			vertices[RegionAttachment.C2A] = color.a;
			vertices[RegionAttachment.U2] = uvs[2];
			vertices[RegionAttachment.V2] = uvs[3];

			vertices[RegionAttachment.C3R] = color.r;
			vertices[RegionAttachment.C3G] = color.g;
			vertices[RegionAttachment.C3B] = color.b;
			vertices[RegionAttachment.C3A] = color.a;
			vertices[RegionAttachment.U3] = uvs[4];
			vertices[RegionAttachment.V3] = uvs[5];

			vertices[RegionAttachment.C4R] = color.r;
			vertices[RegionAttachment.C4G] = color.g;
			vertices[RegionAttachment.C4B] = color.b;
			vertices[RegionAttachment.C4A] = color.a;
			vertices[RegionAttachment.U4] = uvs[6];
			vertices[RegionAttachment.V4] = uvs[7];

			return vertices;
		}

		private computeMeshVertices(slot: Slot, mesh: MeshAttachment, pma: boolean) {
			let skeleton = slot.bone.skeleton;
			let skeletonColor = skeleton.color;
			let slotColor = slot.color;
			let regionColor = mesh.color;
			let alpha = skeletonColor.a * slotColor.a * regionColor.a;
			let multiplier = pma ? alpha : 1;
			let color = this.tempColor;
			color.set(skeletonColor.r * slotColor.r * regionColor.r * multiplier,
					  skeletonColor.g * slotColor.g * regionColor.g * multiplier,
					  skeletonColor.b * slotColor.b * regionColor.b * multiplier,
					  alpha);

			let numVertices = mesh.worldVerticesLength / 2;
			if (this.vertices.length < mesh.worldVerticesLength) {
				this.vertices = Utils.newFloatArray(mesh.worldVerticesLength);
			}
			let vertices = this.vertices;
			mesh.computeWorldVertices(slot, 0, mesh.worldVerticesLength, vertices, 0, SkeletonRenderer.VERTEX_SIZE);

			let uvs = mesh.uvs;
			for (let i = 0, n = numVertices, u = 0, v = 2; i < n; i++) {
				vertices[v++] = color.r;
				vertices[v++] = color.g;
				vertices[v++] = color.b;
				vertices[v++] = color.a;
				vertices[v++] = uvs[u++];
				vertices[v++] = uvs[u++];
				v += 2;
			}

			return vertices;
		}
	}
}
